tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/service/websocket/FailoverClientTest.java

```diff
@@ -0,0 +1,174 @@
+    private static final String MESSAGE = "hi all";
+    private WebSocketRemoteServer firstRemoteServer;
+    private WebSocketRemoteServer secondRemoteServer;
+    private WebSocketRemoteServer thirdRemoteServer;
+
+    @Test(description = "Tests the failover webSocket client by starting the second server in the target URLs.")
+    public void testTextFrameWithSecondServer() throws URISyntaxException, InterruptedException,
+            BallerinaTestException {
+        secondRemoteServer = initiateServer(SECOND_SERVER_PORT);
+        WebSocketTestClient client = initiateClient(URL);
+        sendTextDataAndAssert(client);
+        closeConnection(client, secondRemoteServer);
+    }
+
+    @Test(description = "Tests the failover webSocket client by starting the first server in the target URLs.")
+    public void testBinaryFrameWithFirstServer() throws URISyntaxException, InterruptedException,
+            BallerinaTestException {
+        firstRemoteServer = initiateServer(FIRST_SERVER_PORT);
+        WebSocketTestClient client = initiateClient(URL);
+        sendBinaryDataAndAssert(client);
+        closeConnection(client, firstRemoteServer);
+    }
+
+    @Test(description = "Tests the failover webSocket client by starting the third server in the target URLs.")
+    public void testBinaryFrameWithThirdServer() throws URISyntaxException, InterruptedException,
+            BallerinaTestException {
+        thirdRemoteServer = initiateServer(THIRD_SERVER_PORT);
+        WebSocketTestClient client = initiateClient(URL);
+        sendBinaryDataAndAssert(client);
+        closeConnection(client, thirdRemoteServer);
+    }
+
+    @Test(description = "Tests the failover webSocket client by not starting any of the servers in the targets URLs")
+    public void testFailingFailover() throws URISyntaxException, InterruptedException {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        WebSocketTestClient client = initiateClient(URL);
+        client.setCountDownLatch(countDownLatch);
+        countDownLatch.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        CloseWebSocketFrame closeWebSocketFrame = client.getReceivedCloseFrame();
+        Assert.assertNotNull(closeWebSocketFrame);
+        Assert.assertEquals(closeWebSocketFrame.statusCode(), 1011);
+        Assert.assertTrue(closeWebSocketFrame.reasonText().contains("Unexpected condition"));
+        closeWebSocketFrame.release();
+    }
+
+    @Test(description = "Tests the failover client when getting a handshake timeout")
+    public void testHandshakeTimeout() throws URISyntaxException, InterruptedException, BallerinaTestException {
+        WebSocketTestClient client = initiateClient("ws://localhost:21034");
+        firstRemoteServer = initiateServer(FIRST_SERVER_PORT);
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        countDownLatch.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        sendTextDataAndAssert(client);
```

Comment:
how is timeout checked here?

Reply:
Here, we are waiting until handshake timeout. The timeout error is an IOError. Therefore we need to do the failover.